### PR TITLE
wip:refactor: refactor retry logic of creating pod failed

### PR DESCRIPTION
**Pull request by zhujian7**
**What this PR does / why we need it**:

The [old retry logic](https://github.com/caicloud/cyclone/commit/70d94fdebd540c167b5a592049e3b1135048259f#diff-4b5771587e574593cc8bd245903d6409R80) will make the controller stuck for a long time, delete it.

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @your-reviewer

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by zhujian7** [comment]
/hold 
GC mechanism can only process the latest pod of a stage, the older pods remain in the cluster.

File: pkg/apis/cyclone/v1alpha1/workflow_run.go [hunk 1]
```diff
@@ -190,6 +191,18 @@ type Status struct {
...
+
+	// RetryStatus describes status of retry for a stage.
+	RetryStatus *RetryStatus `json:"retryStatus,omitempty"`
+}
+
+// RetryStatus describes status of retry for a stage which has been
+// executed but not succeeded for some retryable reason.
+type RetryStatus struct {
+	// Times records the execution times, e.g. 1 represents the first time to execute.
+	Times int `json:"times,omitempty"`
```
Comment (supereagle) [inline pkg/apis/cyclone/v1alpha1/workflow_run.go]: I think `counts` is better.

File: pkg/workflow/controller/config.go [hunk 2]
```diff
@@ -109,6 +111,14 @@ type ParallelismConfig struct {
 	SingleWorkflow ParallelismConstraint `json:"single_workflow"`
 }
 
+// RetryConfig configures retry times and duration for retrying to execute a stage.
+type RetryConfig struct {
+	// Times of a stage can be totally retried
+	Times int `json:"times"`
```
Comment (supereagle) [inline pkg/workflow/controller/config.go]: The fields are not clear, what about:

```
type RetryConfig struct {
	// Limit defines the max retry times of one stage execution.
	Limit int `json:"limit"`
	// Timeout defines the retry timeout of one stage execution.
	Timeout time.Duration `json:"Timeout"`
```

**Comment by zhujian7** [comment]
https://github.com/caicloud/cyclone/pull/1528 fixed this problem
/close

**Pull request closed without merging**